### PR TITLE
implemented a recursive descent parser

### PR DESCRIPTION
**Pull request by eyluo**
The parser was based on [this blog post](https://eli.thegreenplace.net/2012/08/02/parsing-expressions-by-precedence-climbing), and supports left-right associativity as well as operation precedence.

### CODE REVIEW

File: ocaml/ast.ml [hunk 1]
```diff
@@ -0,0 +1,18 @@
+open Core
+
```
Comment (hcnelson99) [inline ocaml/ast.ml:2]: If you want you can write something like `module L = Lexer` up here and then use `L` everywhere instead of `Lexer` to make your code shorter.

File: ocaml/ast.ml [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+    match e with
+    | Operator (op, e1, e2) -> 
+        let op_str = 
+            (match op with
+            | Lexer.Pow -> "^"
+            | Lexer.Plus -> "+"
+            | Lexer.Minus -> "-"
+            | Lexer.Divide -> "/"
+            | Lexer.Times -> "*")
+        in String.concat ["(" ; string_of_exp e1 ; op_str ; string_of_exp e2 ; ")"]
```
Comment (hcnelson99) [inline ocaml/ast.ml]: Could use `^` here if you wanted to.

File: ocaml/lexer.mli [hunk 3]
```diff
@@ -11,6 +14,8 @@ type token =
 
 type lexer
 
+val op_info : op -> (prec * assoc)
```
Comment (hcnelson99) [inline ocaml/lexer.mli]: ditto

File: ocaml/lexer.ml [hunk 4]
```diff
@@ -30,6 +34,12 @@ type lexer = {
 let slurp fname = In_channel.with_file ~binary:false fname 
     ~f:(fun ch -> In_channel.input_all ch)
 
+let op_info op =
```
Comment (hcnelson99) [inline ocaml/lexer.ml]: I would put this function in the parser since precedence is a parsing thing, not a lexing thing. This would also let you hide it entirely in the parser implementation and not have it exposed in any interface.

File: ocaml/lexer.mli [hunk 5]
```diff
@@ -1,6 +1,9 @@
 (* Lexer interface. *)
 
-type op = Pow | Plus | Times | Divide | Minus
+type op = Pow | Plus | Minus | Times | Divide
+
+type prec = int
```
Comment (hcnelson99) [inline ocaml/lexer.mli]: Move to parser? (see above)

File: ocaml/lexer.ml [hunk 6]
```diff
@@ -30,6 +34,12 @@ type lexer = {
 let slurp fname = In_channel.with_file ~binary:false fname 
     ~f:(fun ch -> In_channel.input_all ch)
 
+let op_info op = 
+    match op with
+    | Pow -> (3, Right)
+    | Plus | Minus -> (1, Left)
+    | Times | Divide -> (2, Left)
+
```
Comment (hcnelson99) [inline ocaml/lexer.ml:33]: Can't comment on your lexer implementation below since it's not part of this PR, but it looks good. Only comment is I think everywhere you use `!` you don't need the extra parentheses.

File: ocaml/parser.ml [hunk 7]
```diff
@@ -1 +1,49 @@
...
+
+let parse lexer =
+    let rec parse_expr min_prec =
+        let lhs = parse_atom () in
+        let result = parse_expr_prec min_prec lhs in
+        result
+    and parse_expr_prec min_prec lhs = 
+        let tok = Lexer.peek lexer in
+        match tok with
+        | Lexer.Eof | Lexer.IntVal _ -> let () = Lexer.drop lexer in lhs
```
Comment (hcnelson99) [inline ocaml/parser.ml]: Can be `Lexer.drop lexer; lhs`

File: ocaml/parser.ml [hunk 8]
```diff
@@ -1 +1,49 @@
...
+            else 
+                (* Only consume the token if it will be used in the AST *)
+                let () = Lexer.drop lexer in
+                (* For left-associativity, increase precedence to left-group items *)
+                let next_min_prec = 
+                    match assoc with
+                    | Lexer.Left -> prec + 1
+                    | Lexer.Right -> prec
+                in
+                let rhs = (parse_expr next_min_prec) in
```
Comment (hcnelson99) [inline ocaml/parser.ml]: don't need parens

File: ocaml/parser.ml [hunk 9]
```diff
@@ -1 +1,49 @@
...
+                in
+                let rhs = (parse_expr next_min_prec) in 
+                parse_expr_prec min_prec (Ast.Operator(op, lhs, rhs))
+        | _ -> lhs
+    and parse_atom () =
+        let tok = Lexer.pop lexer in
+        match tok with
+        | Lexer.IntVal i -> Ast.IntVal i
+        | Lexer.LParen ->
+            let e = parse_expr (0) in
```
Comment (hcnelson99) [inline ocaml/parser.ml]: don't need parens

**Review by hcnelson99 — COMMENTED** [review]
Looks really good overall

File: ocaml/lexer.ml (hunk 6 above)
Comment (eyluo) [reply ocaml/lexer.ml:33]: I get compilation errors when this happens. I suspect it's because ! takes precedence over ., so it tries to get the `.pos` field of a lexer ref.

Comment (hcnelson99) [reply ocaml/lexer.ml:33]: That's very sad. No reason for precedence to be that way.
